transport/httptransport/libp2p_server.go

```diff
@@ -0,0 +1,591 @@
+// read / write pipe), and waits for the error or completed event to be
+// fired before returning.
+// It returns the state that the transfer is in after being canceled (either
+// completed or errored out).
+func (m *transfersMgr) CancelTransfer(ctx context.Context, id string) (*types.TransferState, error) {
+	// Remove the transfer from the list of active transfers
+	m.transfersLk.Lock()
+	xfer, ok := m.transfers[id]
+	if ok {
+		delete(m.transfers, id)
+	}
+	m.transfersLk.Unlock()
+
+	if !ok {
+		return nil, ErrTransferNotFound
+	}
+
+	// Cancel the transfer
+	return xfer.cancel(ctx)
+}
+
+type MatchFn func(xfer *Libp2pTransfer) (bool, error)
+
+// Matching returns all transfers selected by the match function
+func (m *transfersMgr) Matching(match MatchFn) ([]*Libp2pTransfer, error) {
+	m.transfersLk.RLock()
+	defer m.transfersLk.RUnlock()
+
+	matching := make([]*Libp2pTransfer, 0)
+	for _, xfer := range m.transfers {
+		matches, err := match(xfer)
+		if err != nil {
+			return nil, err
+		}
+		if matches {
+			matching = append(matching, xfer)
+		}
+	}
+
+	return matching, nil
+}
+
+// Adds the transfer to the list of active transfers.
+// It returns a function that fires an event with the current state of the
+// transfer. The function is guaranteed to fire the event after the transfer
+// has been added to the list of active transfers.
+func (m *transfersMgr) add(xfer *Libp2pTransfer) (func(types.TransferState), error) {
+	// Queue up an action that adds the transfer to the map of transfers
+	err := m.enqueueAction(&xferAction{xfer: xfer})
+	if err != nil {
+		return nil, err
+	}
+
+	// fireEvent is called when the state changes (eg data is sent, or there's an error)
+	fireEvent := func(st types.TransferState) {
+		// Queue up an action that fires an event with the current transfer state.
+		// Note: enqueueAction returns an error only if the context is cancelled,
+		// which we can safely ignore.
+		_ = m.enqueueAction(&xferAction{xfer: xfer, transferState: &st}) //nolint:errcheck
+	}
+
+	return fireEvent, nil
+}
+
+type xferAction struct {
+	xfer          *Libp2pTransfer
+	transferState *types.TransferState
+}
+
+// enqueueAction adds an action to the queue, or returns context.Canceled if
+// the transfersMgr has been shut down
+func (m *transfersMgr) enqueueAction(evt *xferAction) error {
+	select {
+	case <-m.ctx.Done():
+		return m.ctx.Err()
+	case m.actions <- evt:
+		return nil
+	}
+}
+
+// run processes actions on the queue until the context is cancelled
+func (m *transfersMgr) run(ctx context.Context) {
+	defer close(m.done)
+
+	m.ctx = ctx
+
+	processAction := func(action *xferAction) {
+		xfer := action.xfer
+		// There are two types of action:
+
+		// 1. Add a new transfer to the list of active transfers
+		if action.transferState == nil {
+			m.transfersLk.Lock()
+			if existing, ok := m.transfers[xfer.ID]; ok {
+				// Cancel any existing transfer with the same id
```

Comment:
This is also racy as we could probably emit ongoing events for the new transfer before emitting the cancellation/failure of the previous transfer and the events will have the same `ID`.  

I think the best course of action here is to initiate the sendcar/serve content for a given ID  only after all previous transfers for that ID have been cancelled **and** cancel has returned so we don't race the new transfer events with the previous transfer events.

We should also emit some sort of Restart event if it helps Estuary/the client whenever we cancel a previous transfer before initiating a new one. What if they give up after listening to the previous transfer ? But, leave this decision to you.

Reply:
The other solution is to the put a unique monotonically increasing "sessionID" on each transfer request that is emitted along with the event and have clients ignore events with lower session IDs. But that will need clients to keep some state.  I think it's best to avoid this and just go ahead with the "cancel everything before serving a new request" solution.